byron/ledger/executable-spec/test/Ledger/Upiec/Properties.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Ledger.Upiec.Properties
+  ( noProtVerChange
+  , protVerChangeAdopt
+  , protVerChangeSameComponents
+  , protVerChangeEmptyComponents
+  ) where
+
+import           Control.Lens (_1, _2, _3, _4, _5, _6, _7, _8, _9, (^.))
+import           Control.State.Transition (applySTS, TRC(..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import           Ledger.Core (Slot)
+import           Ledger.Update (UpId)
+import qualified Ledger.Update.Generators as G
+import           Hedgehog
+import           Ledger.Update (PVBUMP, UPIEC, ProtVer, PParams)
+
+
+-- Property #1 for the UPIEC STS
+--
+-- If PVBUMP transitions into a state with the same protocol version,
+-- the UPIEC STS results in the same state it started in.
+noProtVerChange :: Property
+noProtVerChange = property $ do
+  jc <- forAll G.noProtVerChangeJC
+  let (_, st, _) = jc
+  case applySTS @UPIEC (TRC jc) of
+    Right st' -> st === st'
+    Left _    -> failure
+
+-- Property #2 for the UPIEC STS
```

Comment:
I'm a bit weary of encoding the properties in this way: we're mimicking the rules to a high level of detail, which means that if we change something in the rules we need to change a lot of implementation details in the tests. For instance, if we would replace `PVBUMP` by something else but keep the same high level properties of `UPIEC` we would have to re-write this test, even though there are no changes in the the observable behavior of `UPIEC`.

Reply:
the makes sense to me - that the exec spec reflects the STS hierarchy (to allow for re-use). Damian, perhaps to address your concern we could make this layering of STS's more explicitly composable in order lower the cost of change? Or what alternatives do you have in mind?

Reply:
> perhaps to address your concern we could make this layering of STS's more explicitly composable in order lower the cost of change?

Hmmm, I don't follow this, sorry. Can you give an example? :) 

>  Or what alternatives do you have in mind?

Well basically [this](https://github.com/input-output-hk/cardano-ledger-specs/pull/527#discussion_r289739926).

Reply:
@dnadales , I'm all for discussing what properties for STSes should be. The way I've been proposing properties so far is to write them down in an issue (for example, for the `UPIEC` STS in issue #512) and then ask for feedback. If you don't mind I'll @-mention you in the future so that properties are brought to your attention before I start implementing them.

In general, regardless of a testing method used (e.g., unit testing, property-based testing) once you change an implementation you'll have to change tests as well. There's no surprise there. If you change a lot of details in the implementation of rules, then I'd expect to see roughly the same amount of work that has to go to updating the corresponding tests.

Do you have a concrete proposal what a property should look like for `UPIEC`?

Reply:
>  If you don't mind I'll @-mention you in the future so that properties are brought to your attention before I start implementing them.

Sure. FWIW, I read the properties you stated, but I couldn't know in advance how they will be implemented.

> In general, regardless of a testing method used (e.g., unit testing, property-based testing) once you change an implementation you'll have to change tests as well.

This is not true in general. If you change the implementation but do not change the observable properties related to functional requirements of it, e.g. think of performance optimizations, you don't necessarily have to change the tests. 

And even if you change functional properties, the tests should be changed but not in a way that mimics the change in the actual implementation. Whenever possible we should try to separate both.

> If you change a lot of details in the implementation of rules, then I'd expect to see roughly the same amount of work that has to go to updating the corresponding tests.

This also depends on the nature of the properties. If the properties are high level: e.g. money is constant in the system, then no matter what change we make in the UTxO TS, we shouldn't need to change the tests that much.

> Do you have a concrete proposal what a property should look like for UPIEC?

￼Just the other case for Property `#1`, when the protocol version changes.

Reply:
I'd argue it's a bit unfair to compare a low-level STS `UPIEC` with a high-level STS for UTxO. I'm expecting properties will get more interesting and high-level as I get to testing higher-level STSes.

> Just the other case for Property #1, when the protocol version changes.

So properties no. 2 through 4?